lib/ui/semantics.dart

```diff
@@ -641,6 +641,164 @@ class SemanticsFlag {
   }
 }
 
+/// A string that carries a list of [StringAttribute]s.
+class AttributedString extends NativeFieldWrapperClass2 {
+  /// Creates a attributed string.
+  ///
+  /// The [TextRange] in the [attributes] must be inside the length of the
+  /// [string].
+  AttributedString(
+    this.string, {
+    this.attributes,
+  }) : assert(() {
+         if (attributes != null) {
+           for (final StringAttribute attribute in attributes) {
+             assert(
+               string.length >= attribute.range.start &&
+               string.length >= attribute.range.end,
+               'The range in $attribute is outside of the string $string',
+             );
+           }
+         }
+         return true;
+       }()) {
+    _initAttributedString(this, string, attributes);
+  }
+
+  /// The plain string stored in the attributed string.
+  final String string;
+
+  // The attributes this string carries.
+  //
+  // The list must not be modified after this string is created.
+  final List<StringAttribute>? attributes;
+
+  void _initAttributedString(
+    AttributedString instance,
+    String string,
+    List<StringAttribute>? attributes,
+  ) native 'AttributedString_initAttributedString';
+
+  @override
+  AttributedString operator +(AttributedString other) {
+    if (string.isEmpty) {
+      return other;
+    }
+    if (other.string.isEmpty) {
+      return this;
+    }
```

Comment:
I'm confused about what I would expect this use case to do:

```dart
AttributedString a = AttributedString('asdf', [StringAttribute(...), StringAttribute(...)]);
AttributedString b = a + AttributedString('', [StringAttribute(...)]);
```

Why wouldn't these just concat the attributes?

If this is invalid, we should probably make it invalid in the ctor to have an empty string.

Reply:
I can make an assert to avoid empty string with non empty attributes. The StringAttribute would have empty range anyway if the string is empty, in this case, concat attribute with empty range does not do anything but wasting memory

Reply:
If they're blocked from construction we can just change these to asserts

Reply:
Oh I meant, it is allow to have empty AttributedString with empty string, for non null property in SemanticsConfiguration and SemanticsNode on the framework side, unless we want to change that.

Reply:
Asserting that when string is empty, attributes are empty as well sounds good to me and then we keep this logic?